### PR TITLE
Add support for multihost pmaps.

### PR DESCRIPTION
**Pull request by skye**
All participating hosts are assumed to be running the same pmap code. Conceptually, this can be considered a single pmap over an array sharded on its leading pmapped dimension across the hosts. Each host passes its input shard to its pmapped function call, which returns the corresponding output shard (i.e. an array of the same leading dimension size). However, any collective operations will be run across the entire "global" array.

If the `devices` argument to pmap is None, the pmap is assumed to be running across all hosts visible to XLA (as returned by jax.host_count()). Each host can pass in an input array of leading dimension size equal to or less than the number of devices local to that host. Note that this doesn't change the current behavior for single-host platforms. If `devices` are specified, the participating hosts are dictated by the devices' host_ids, and each host must pass in an input array of leading dim size equal to the number of local participating devices.

Implementation-wise, each host independently compiles the computation, which we assume yields the same executable on all hosts (follow-up work will add more error checking). The hosts must know the global axis size of the sharded array, e.g. to provide the correct replica count to XLA. This is equal to the length of `devices` if specified, but if not, pmap is recursively called (with `devices` specified) to use `psum` to compute the global axis size.

### CODE REVIEW

File: jax/interpreters/pxla.py [hunk 1]
```diff
@@ -166,31 +167,36 @@ def replica_groups(nrep, mesh_spec, mesh_axes):
...
 ### the main pmap machinery lowers SPMD jaxprs to multi-replica XLA computations
 
-def compile_replicated(jaxpr, backend, axis_name, axis_size, devices, consts,
-                       *abstract_args):
+def compile_replicated(jaxpr, backend, axis_name, axis_size, global_axis_size,
+                       devices, consts, *abstract_args):
+  jaxpr_replicas = xla.jaxpr_replicas(jaxpr)
+  num_local_replicas = axis_size * jaxpr_replicas
+  num_replicas = global_axis_size * jaxpr_replicas
+  logging.info(
```
Comment (hawkinsp) [inline jax/interpreters/pxla.py]: `info` feels a bit verbose. `vlog`?

File: jax/interpreters/pxla.py [hunk 2]
```diff
@@ -461,13 +474,44 @@ def dynamic_fun(dummy, *args):
...
+  if devices is None:
+    # Default behavior is running on all devices/hosts.
+    return xb.host_count() > 1
+  else:
+    return len(set(d.host_id for d in devices)) > 1
+
+_get_global_axis_size_pmapped = None
+
+def _get_global_axis_size(local_axis_size):
+  """Uses pmap to sum `local_axis_size` across all hosts.
```
Comment (hawkinsp) [inline jax/interpreters/pxla.py]: While this approach is cunning, don't we know the global topology from the low-level runtime stack? Can't we use that here instead?

In particular the module dependency structure seems a bit odd to me with this design.

It also feels strange to me that we have to special case multihost computations here; it feels to me like the same code should work in both single host and multihost cases.

**Review by hawkinsp — COMMENTED** [review]
You should ideally get a review from Matt if he can.

File: jax/interpreters/pxla.py (hunk 1 above)
Comment (skye) [reply jax/interpreters/pxla.py]: Done.

File: jax/interpreters/pxla.py (hunk 2 above)
Comment (skye) [reply jax/interpreters/pxla.py]: As discussed offline, I got rid of this for now and instead require that multihost pmaps run on all devices or have `devices` specified.

File: jax/api.py [hunk 3]
```diff
@@ -51,7 +51,7 @@
 from .util import (unzip2, unzip3, curry, partial, safe_map, safe_zip,
                    WrapHashably, Hashable, prod, split_list)
 from .lib.xla_bridge import (canonicalize_dtype, device_count,
-                             local_device_count, devices, host_id)
+                             local_device_count, devices, host_id, host_count)
```
[the remark this answers is not in the document]
Comment (skye) [reply jax/api.py]: These are imported to bring them into the `jax` namespace, e.g. so you can call `jax.host_count()`. I would prefer we dealt with this in a different way, but haven't gotten around to trying anything...

**Review by hawkinsp — APPROVED** [review]

**Review by mattjj — APPROVED** [review]
LGTM! Clear explanations, simple `global_axis_size` implementation, plus logging and error checking. Thanks for adding this!

**Comment by skye** [comment]
Thanks Matt!

FYI I reworked the multi-host pmap example in the pmap docstring based on offline suggestions from @necula01. I'm gonna submit now as-is, but if anyone has further comments, I can iterate in subsequent commits.
